File: algorithms/sort/quick.cpp
#include<bits/stdc++.h>
using namespace std;

int generate_random(int low, int high){
  return low + rand()%(high-low+1);
}

int partition(vector<int> &arr, int low, int high){
  int p = generate_random(low, high);
  swap(arr[low], arr[p]);
  int pivot = arr[low];
  int j = low + 1;

  for(int i = low+1; i <= high; i++){
    if(arr[i] < pivot) swap(arr[i], arr[j++]);
  }

  swap(arr[low], arr[j-1]);
  return j-1;
}

void quick_sort(vector<int> &arr, int low, int high){
  if(low >= high) return;
  int p = partition(arr, low, high);
  quick_sort(arr, low, p-1);
  quick_sort(arr, p+1, high);
}

int main(){
  vector<int> arr = {4, 5, 1, 3, 2};
  quick_sort(arr, 0, 4);
  for (int i = 0; i < arr.size(); i++){
    cout << arr[i] << " ";
  }  
} 

File: algorithms/sort/quick.py
from random import randint

def random_index(left, right):
  return randint(left, right)

def partition(list, left, right):
  pivot = random_index(left, right)
  list[left], list[pivot] = list[pivot], list[left]
  index = left + 1

  for i in range(left + 1, right + 1):
    if list[i] < list[left]:
      list[i], list[index] = list[index], list[i]
      index += 1

  pivot = index - 1
  list[left], list[pivot] = list[pivot], list[left]
  return pivot

def quick_sort(list, left, right):
  if left < right:
    pivot = partition(list, left, right)
    quick_sort(list, left, pivot - 1)
    quick_sort(list, pivot + 1, right)

list = [4, 5, 9, 8, 7, 3, 2, 1, 6]
quick_sort(list, 0, len(list)-1)
print(list)
